src/collections/vector.h

```diff
@@ -54,19 +90,31 @@ inline bool Vec<T>::is_empty() {
 
 template <typename T>
 void Vec<T>::push(T data) {
-  T* new_buf = new T[this->len + 1];
-  std::copy(this->buf, this->buf + this->len, new_buf);
-  new_buf[this->len] = data;
-
-  delete[] this->buf;
-
-  this->buf = new_buf;
-  this->len++;
+  size_t total = this->len + 1;
+  if (total > this->_capacity) {
+    if (this->_capacity == 0) {
+      this->_capacity = 1;
+    } else {
+      this->_capacity = this->_capacity * 2;
+    }
+    reallocate();
+  }
+  buf[this->len++] = data;
 }
 
 template <typename T>
 std::optional<T> Vec<T>::pop() {
-  if (this->len > 0) return this->buf[this->len-- - 1];
+  if (this->len > 0) {
```

Comment:
please add some comments here notifying that this function is not exception-safe.